[scroll-anchoring] Implement anchor node selection algorithm

nmoucht opened the pull request:
#### 0cd807347378c72b30ee428774daeaf2b04cff07
<pre>
[scroll-anchoring] Implement anchor node selection algorithm
<a href="https://bugs.webkit.org/show_bug.cgi?id=261628">https://bugs.webkit.org/show_bug.cgi?id=261628</a>
rdar://115583079

Reviewed by Simon Fraser.

Implement anchor node selection algorithm outlined in the spec.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/descend-into-container-with-float-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/exclude-fixed-position-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/exclude-sticky-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/position-change-heuristic-ib-split-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/subtree-exclusion-expected.txt:
* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::anchorElementForPriorityCandidate):
(WebCore::elementIsScrollableArea):
(WebCore::ScrollAnchoringController::examinePriorityCandidate):
(WebCore::ScrollAnchoringController::findPriorityCandidate):
(WebCore::ScrollAnchoringController::examineCandidate):
(WebCore::operator&lt;&lt;):
(WebCore::ScrollAnchoringController::findAnchorElementRecursive):
(WebCore::ScrollAnchoringController::chooseAnchorElement):
(WebCore::ScrollAnchoringController::updateAnchorElement):
(WebCore::ScrollAnchoringController::adjustScrollPositionForAnchoring):
* Source/WebCore/page/scrolling/ScrollAnchoringController.h:

Canonical link: <a href="https://commits.webkit.org/268348@main">https://commits.webkit.org/268348@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43940edb75a6ef190ce9141b241db3e4c773c4b4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19401 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19819 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20417 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21288 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18138 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19634 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23084 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19939 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19761 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19615 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19654 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/16860 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22147 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16840 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17648 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23961 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17902 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17823 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21931 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18409 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15596 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17553 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/17484 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21910 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2379 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18241 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/accessibility/visible-character-range-width-changes.html:
```diff
@@ -1,3 +1,4 @@
+<!-- webkit-test-runner [ CSSScrollAnchoringEnabled=false ] -->
```
It would be preferable to use the CSS property, although I would guess it's not implemented yet.

nt1m replied:
I guess your commit does implement part of it, so it might be possible here.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -50,35 +53,138 @@ ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
-    // (specifically excluded subtrees and priority nodes)
-    // https://drafts.csswg.org/css-scroll-anchoring/#anchor-node-selection
+    while (element) {
+        if (auto renderer = element->renderBox()) {
```
Is there a reason to use `renderBox()` and not `renderer()`?

nmoucht replied:
Right should be `renderer()`.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -50,35 +53,138 @@ ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
+            LOG_WITH_STREAM(ScrollAnchoring, stream << "ScrollAnchoringController::examinePriorityCandidate() for element: " << *iterElement << " element scrolling id: " << ((element->renderBox() && element->renderBox()->layer() && element->renderBox()->layer()->scrollableArea()) ? (element->renderBox()->layer()->scrollableArea()->scrollingNodeID()) : 0) << " current Frame: " << m_frameView << "f ...
 
-    Element* candidateAnchor = nullptr;
+            if (auto renderer = element->renderBox()) {
```
Is there a reason to use `renderBox()` and not `renderer()`?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -50,35 +53,138 @@ ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
+        auto boxRect = renderer->absoluteBoundingBoxRect();
+        if (!boxRect.width() || !boxRect.height())
+            return CandidateExaminationResult::Skip;
+        if (layoutViewport.contains(boxRect))
```
Need tests with various types of zooming to ensure that `absoluteBoundingBoxRect` and `layoutViewport` share the same coordinate space in those scenarios.

nmoucht replied:
This will also need some changing in https://github.com/WebKit/WebKit/pull/17934 as we want to use the visible area for the owning scroller.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -50,35 +53,138 @@ ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
-        return;
+bool ScrollAnchoringController::examinePriorityCandidate(Element* element)
+{
+    if (auto* candidateElement = anchorElementForPriorityCandidate(element)) {
```
Prefer an early return.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -50,35 +53,138 @@ ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
+{
+    if (auto* candidateElement = anchorElementForPriorityCandidate(element)) {
+        LOG_WITH_STREAM(ScrollAnchoring, stream << "ScrollAnchoringController::examinePriorityCandidate()");
+        auto* iterElement = candidateElement;
```
Our smart pointer rules say that iterElement should be a RefPtr.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -50,35 +53,138 @@ ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
+
+    findPriorityCandidate(document);
+
+    Element* anchorElement = nullptr;
```
Use a RefPtr.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -50,35 +53,138 @@ ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
+    if (!m_anchorElement)
+        anchorElement = findAnchorElementRecursive(document.documentElement());
+
+    if (!anchorElement)
```
A bit confused about why we don't change `m_anchorElement` if we fail to find a new one.

nmoucht replied:
We only call `chooseAnchorElement` if `m_anchorElement` is null.

smfr replied:
I feel like nesting the ifs would make this more clear.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.h:
```diff
@@ -32,20 +32,31 @@ namespace WebCore {
 private:
+    Element* findAnchorElementRecursive(Element*);
+    CandidateExaminationResult examineCandidate(Element&);
+    bool examinePriorityCandidate(Element*);
```
You should name the functions to make it clear what the bool return values mean.

Can these be static?

nmoucht replied:
Changed these to static.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -43,42 +47,143 @@ void ScrollAnchoringController::invalidateAnchorElement()
+    if (!candidateElement)
+        return nullptr;
+
+    LOG_WITH_STREAM(ScrollAnchoring, stream << "ScrollAnchoringController::examinePriorityCandidate()");
```
Logging doesn't match the new function name.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -43,42 +47,143 @@ void ScrollAnchoringController::invalidateAnchorElement()
-    auto* document = m_frameView.frame().document();
-    if (!document)
-        return;
+static RefPtr<Element> viablePriorityCandidateForElement(Element* element)
```
If you have a static function that's just used at one call site, consider just making it a lambda.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -43,42 +47,143 @@ void ScrollAnchoringController::invalidateAnchorElement()
+            if (auto* anchorElement = findAnchorElementRecursive(&child))
+                return anchorElement;
         }
     }
```
Add a break at the end of this case statement, so if we add new ones in future we don't get fall-through by mistake.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -87,8 +192,12 @@ void ScrollAnchoringController::updateAnchorElement()
 
-    if (!m_anchorElement || !m_anchorElement->renderBox()) {
-        chooseAnchorElement();
+    auto* document = m_frameView.frame().document();
```
RefPtr

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-anchoring] Implement anchor node selection algorithm
https://bugs.webkit.org/show_bug.cgi?id=261628
rdar://115583079

Reviewed by Simon Fraser.

Implement anchor node selection algorithm outlined in the spec.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/descend-into-container-with-float-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/exclude-fixed-position-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/exclude-sticky-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/position-change-heuristic-ib-split-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/subtree-exclusion-expected.txt:
* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::anchorElementForPriorityCandidate):
(WebCore::elementIsScrollableArea):
(WebCore::ScrollAnchoringController::examinePriorityCandidate):
(WebCore::ScrollAnchoringController::findPriorityCandidate):
(WebCore::ScrollAnchoringController::examineCandidate):
(WebCore::operator<<):
(WebCore::ScrollAnchoringController::findAnchorElementRecursive):
(WebCore::ScrollAnchoringController::chooseAnchorElement):
(WebCore::ScrollAnchoringController::updateAnchorElement):
(WebCore::ScrollAnchoringController::adjustScrollPositionForAnchoring):
* Source/WebCore/page/scrolling/ScrollAnchoringController.h:

Canonical link: https://commits.webkit.org/268348@main
